### DISCUSSION TITLE
How do multi-center attachment points work internally?

### DISCUSSION THREAD

**Discussion by kienerj** [rdkit/rdkit#6399] [Q&A] [answered]
I'm trying to figure out of the multi-center attachment points work in rdkit:

```python
Chem.MolFromMolBlock("""
  ChemDraw05262308532D

  0  0  0     0  0              0 V3000
M  V30 BEGIN CTAB
M  V30 COUNTS 9 8 0 0 0
M  V30 BEGIN ATOM
M  V30 1 C -0.238139 0.481238 0.000000 0
M  V30 2 C -0.952612 0.068742 0.000000 0
M  V30 3 C -0.952612 -0.756248 0.000000 0
M  V30 4 N -0.238139 -1.168744 0.000000 0
M  V30 5 C 0.476282 -0.756248 0.000000 0
M  V30 6 C 0.476282 0.068742 0.000000 0
M  V30 7 O 0.238142 0.756244 0.000000 0
M  V30 8 C 0.952612 1.168744 0.000000 0
M  V30 9 * 0.238142 -0.068756 0.000000 0
M  V30 END ATOM
M  V30 BEGIN BOND
M  V30 1 1 1 2
M  V30 2 2 2 3
M  V30 3 1 3 4
M  V30 4 2 4 5
M  V30 5 1 5 6
M  V30 6 2 1 6
M  V30 7 1 7 8
M  V30 8 1 7 9 ENDPTS=(3 6 1 5) ATTACH=ANY
M  V30 END BOND
M  V30 END CTAB
M  END
""")
```

![image](https://github.com/rdkit/rdkit/assets/1567309/60c51672-c674-4d9d-9c9a-a23b31fde795)

I can see that the ENDPTS is a property of the bond and the bond goes to a dummy atom (*). However one can also see above that the atom indexes don't match so how does RDKit map from the "molfile atom indexes" to the rdkit atom indexes? Or does this work entirely differently and ENDPTS isn't really taken into account?

**Comment by DavidACosgrove** [comment, marked as answer]
Hi Joos,
The 3 is the number of atoms in the feature, the others are the indices of
the atoms associated with the dummy atom, counting from 1. RDKit counts
from zero, obviously.
Dave


On Fri, 26 May 2023 at 08:03, Joos Kiener ***@***.***> wrote:

> I'm trying to figure out of the multi-center attachment points work in
> rdkit:
>
> Chem.MolFromMolBlock("""  ChemDraw05262308532D  0  0  0     0  0              0 V3000M  V30 BEGIN CTABM  V30 COUNTS 9 8 0 0 0M  V30 BEGIN ATOMM  V30 1 C -0.238139 0.481238 0.000000 0M  V30 2 C -0.952612 0.068742 0.000000 0M  V30 3 C -0.952612 -0.756248 0.000000 0M  V30 4 N -0.238139 -1.168744 0.000000 0M  V30 5 C 0.476282 -0.756248 0.000000 0M  V30 6 C 0.476282 0.068742 0.000000 0M  V30 7 O 0.238142 0.756244 0.000000 0M  V30 8 C 0.952612 1.168744 0.000000 0M  V30 9 * 0.238142 -0.068756 0.000000 0M  V30 END ATOMM  V30 BEGIN BONDM  V30 1 1 1 2M  V30 2 2 2 3M  V30 3 1 3 4M  V30 4 2 4 5M  V30 5 1 5 6M  V30 6 2 1 6M  V30 7 1 7 8M  V30 8 1 7 9 ENDPTS=(3 6 1 5) ATTACH=ANYM  V30 END BONDM  V30 END CTABM  END""")
>
> [image: image]
> <https://user-images.githubusercontent.com/1567309/241143418-60c51672-c674-4d9d-9c9a-a23b31fde795.png>
>
> I can see that the ENDPTS is a property of the bond and the bond goes to a
> dummy atom (*). However one can also see above that the atom indexes don't
> match so how does RDKit map from the "molfile atom indexes" to the rdkit
> atom indexes? Or does this work entirely differently and ENDPTS isn't
> really taken into account?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/rdkit/rdkit/discussions/6399>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACGF2FXVTLU36VSASVUNYPDXIBIUXANCNFSM6AAAAAAYP2EZS4>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>
-- 
David Cosgrove
Freelance computational chemistry and chemoinformatics developer
http://cozchemix.co.uk

**Reply by kienerj** [reply]
Thanks. got it. the atom count, eg. the 3 is what confused me.
